src/headerUtils.js

```diff
@@ -112,6 +112,69 @@ module.exports = function(env) {
     return k(s, params[0]);
   };
 
+  // `mapData` maps a function over an array much like the `map`
+  // function. It differs in that the use of `mapData` signals to the
+  // language that the random choices in each `obsFn` are
+  // conditionally independent given the random choices made before
+  // `mapData`.
+
+  // The way this information is used will be coroutine specific. When
+  // the current coroutine doesn't provide specific handling the
+  // behavior is equivalent to regular `map`.
+
+  // This is still somewhat experimental. The interface may change in
+  // the future.
+
+  function mapData(s, k, a, opts, obsFn) {
+    opts = opts || {};
+
+    var data = opts.data;
+    if (!_.isArray(data)) {
+      throw new Error('mapData: No data given.');
+    }
+
+    var batchSize = _.has(opts, 'batchSize') ? opts.batchSize : data.length;
+    if (batchSize <= 0 || batchSize > data.length) {
+      throw new Error('mapData: Invalid batchSize.');
+    }
+
+    // Query the coroutine to determine the subset of the data to map
+    // over.
+    var ix = env.coroutine.mapDataFetch ?
+        env.coroutine.mapDataFetch(data, batchSize, a) :
+        // The empty array stands for all indices, in order. i.e.
+        // `_.range(data.length)`
+        [];
```

Comment:
I wonder if we should use another value as a placeholder for all indices. As it stands, there is no way for the coroutine to actually indicate the empty array (which wouldn't be an entirely implausible thing to do, e.g. to initialize inference without conditioning on the data). Maybe use `null` instead?